### PR TITLE
Enable incremental draft generation for all users

### PR DESCRIPTION
**Pull request by Nateowami**
~Labeled `do not merge` because it's waiting on the fixes in #4070.~

Labeled `testing not required` because it's enabling a feature that's already been fully tested.

- - -
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/4082)

### CODE REVIEW

**Comment by marksvc** [comment]
This is reviewable in [Devin Review](https://app.devin.ai/review/sillsdev/web-xforge/pull/4082).

**Review by marksvc — APPROVED** [review]
> @marksvc reviewed 3 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/4082)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on Nateowami).

**Comment by marksvc** [comment]
Not merging yet to see if e2e failure is real or flaky.

**Comment by Nateowami** [comment]
@marksvc It's real. We have a test that exercises the old drafting stepper and one for the new drafting stepper. I'll work on it.

**Comment by marksvc** [comment]
If helpful, AI suggests a fix might be
```diff
+++ b/src/SIL.XForge.Scripture/ClientApp/e2e/workflows/partial-book-drafting.ts
@@ -2,6 +2,7 @@ import { expect } from 'npm:@playwright/test';
 import { Locator, Page } from 'npm:playwright';
 import { E2E_SYNC_DEFAULT_TIMEOUT, preset, ScreenshotContext } from '../e2e-globals.ts';
 import {
+  enableDeveloperMode,
   enableDraftingOnProjectAsServalAdmin,
   freshlyConnectProject,
   getNewBrowserForSideWork,
@@ -52,6 +53,9 @@ export async function partialBookDrafting(
 
   await freshlyConnectProject(page, TARGET_PROJECT_SHORT_NAME);
 
+  // The wizard's echo/fast-training engine settings step only appears in developer mode.
+  await enableDeveloperMode(page, { closeMenu: true });
+
   await user.click(page.getByRole('link', { name: 'Generate draft' }));
   await expect(page.getByRole('heading', { name: 'Generate translation drafts' })).toBeVisible();
```

**Comment by Nateowami** [comment]
@marksvc Unfortunately fixing the tests is proving to be difficult. We have one basic generate draft test, and when I added partial book drafting I created one that exercises that. I'm not trying to create an overarching test that exercises partial book drafting, but also exercises everything the original generate draft test tested. And now I'm running into weird performance issues that I think are probably caused by an actual bug. For now I'm going to keep investigating and trying to figure out what's going on.

**Comment by Nateowami** [comment]
@marksvc 

I've updated the e2e tests. When I originally implemented partial book drafting I added a new e2e test specifically for it. It focused on partial book drafting and didn't draft a whole book. In this change I retired the old generate_draft test, and then renamed the partial book drafting test to generate_draft. Then in a separate commit the test is updated to also draft a whole book. The commits are separate to make the review easier (otherwise the diff will compare the old generate_draft with the partial book drafting test + edits). I also re-characterized the tests. The commit history might make it look like I re-characterized them before the final changes were made, but it's not entirely chronological, and I actually did it in the correct order, so the results are valid.

Also updated the localized screenshots workflow.

**Comment by marksvc** [comment]
*[`src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts` line 182 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0dNq4t9uXVvz-L6rxr:-P0dNq4u75Fk0yuoIMgG:bqc97ro) ([raw file](https://github.com/sillsdev/web-xforge/blob/ffe2864f04f62357d5e530398a99ea511d0b65ae/src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts#L182)):*
> drafted

I think that what is meant here is that Obadiah *has chapter content in each chapter in the source*, and so that is why it is not in the list of partial books.  If so, could you reword this sentence so it does not use "draft"?

If what you mean is that *content for all chapters of Obadiah will be made by AI*, or *content for all chapters of Obadiah has been made by AI*, then maybe we can pair "draft" with "generation" to clarify that.

**Review by marksvc — COMMENTED** [review]
Thank you for the heads up about how you did the commits, which made it easier to review. 

> @marksvc reviewed 5 files and all commit messages, and made 5 comments.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-:-P0dFfTF-QmDXonkKhmA:b-8xv9of)* status: all files reviewed, 5 unresolved discussions (waiting on Nateowami).

___
*[`src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts` line 312 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0dRyT15aLmKNoz30Im:-P0dRyT15aLmKNoz30In:b-rh4vwi) ([raw file](https://github.com/sillsdev/web-xforge/blob/ffe2864f04f62357d5e530398a99ea511d0b65ae/src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts#L312)):*
>  * Choose formatting options, ... apply its first drafted chapter ...
>  * Because that chapter then has content, the import wizard's overwrite confirmation step is guaranteed to appear in
>  * importDraft.
> ...
> await user.click(page.getByRole('button', { name: 'Overwrite chapter' }));

These don't seem to fit together. 

The code line makes me think the project already has content and so we _will_ get an Overwrite prompt that we need to click. And that's okay. But maybe the function comment could use some wording improvement then. Presumably we are benefiting from testing the apply-one-chapter capability, and getting the screenshot.
___
*[`src/SIL.XForge.Scripture/ClientApp/e2e/test_characterization.json` line 2 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0dJJpS2h2U3rd28asL:-P0dJJpS2h2U3rd28asM:b-6np5t) ([raw file](https://github.com/sillsdev/web-xforge/blob/ffe2864f04f62357d5e530398a99ea511d0b65ae/src/SIL.XForge.Scripture/ClientApp/e2e/test_characterization.json#L2)):*
> ```JSON
> {
>   "smoke_tests": {
> ```

I am aware that the test characterization is build with math and care. But not uncommonly it seems to me that the E2E tests needed to retry more. 
___
*[`src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts` line 36 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0dKnSN33_wcZXwUY_N:-P0dKnSN33_wcZXwUY_O:b-rx6uba) ([raw file](https://github.com/sillsdev/web-xforge/blob/ffe2864f04f62357d5e530398a99ea511d0b65ae/src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts#L36)):*
> ```XML
> // A book drafted in full, using the default chapter selection. It has only one chapter, so the wizard can never default
> // it to a subset of chapters, whatever the target project contains.
> const WHOLE_BOOK_NAME = 'Obadiah';
> ```

Obadiah, with only one chapter, sounds like a good corner case to test *in addition to* the usual scenario of a whole book which has multiple chapters, don't you think? Are you not choosing a multi-chapter book for WHOLE_BOOK_NAME because of a limitation described in the comment above?
___
*[`src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts` line 202 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0dPu1v3Gs4-07s82ay:-P0dPu1v3Gs4-07s82az:b-itythp) ([raw file](https://github.com/sillsdev/web-xforge/blob/ffe2864f04f62357d5e530398a99ea511d0b65ae/src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts#L202)):*
> ```XML
>   }
> 
>   // Selecting a book in the target project automatically selects it in each reference project.
> ```

I am so glad for the various comments you have been improving the explanations and clarity of. Thank you.

**Review by Nateowami — COMMENTED** [review]
> @Nateowami made 4 comments.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/4082)* status: 6 of 7 files reviewed, 5 unresolved discussions (waiting on marksvc).

___
*[`src/SIL.XForge.Scripture/ClientApp/e2e/test_characterization.json` line 2 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0dJJpS2h2U3rd28asL:-P0eBMx42NPIslbqzVSf:b-c9eh3i) ([raw file](https://github.com/sillsdev/web-xforge/blob/ffe2864f04f62357d5e530398a99ea511d0b65ae/src/SIL.XForge.Scripture/ClientApp/e2e/test_characterization.json#L2)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

I am aware that the test characterization is build with math and care. But not uncommonly it seems to me that the E2E tests needed to retry more.
</blockquote></details>

It really serves two purposes:
1. Force us to run a test enough times that we've proven it's somewhat reliable. In practice this is probably the more valuable purpose, even though it wasn't the originally intended purpose. It has definitely stopped me from pushing a test I thought was good.
2. Re-run tests on CI when they fail until it's improbable that they failed for a reason that was present when they were originally characterized. What that means is that when a test fails it's nearly always due to a factor not present when the test was being characterized (machine performance, network speed, code changes, etc). Of course there are numerous ways this could happen without the application having had a true regression.

I've re-characterized generate_draft again.
___
*[`src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts` line 36 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0dKnSN33_wcZXwUY_N:-P0eDpCW8iZJ8ynl4aKE:bo0k81o) ([raw file](https://github.com/sillsdev/web-xforge/blob/ffe2864f04f62357d5e530398a99ea511d0b65ae/src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts#L36)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

Obadiah, with only one chapter, sounds like a good corner case to test *in addition to* the usual scenario of a whole book which has multiple chapters, don't you think? Are you not choosing a multi-chapter book for WHOLE_BOOK_NAME because of a limitation described in the comment above?
</blockquote></details>

I've used Obadiah I lot for testing due to being short, so when using fast training it's much quicker to draft than a longer book. But for echo it doesn't make a difference. Replaced with Daniel, since it has 12 chapters, so it is a book that is eligible for partial book drafting.
___
*[`src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts` line 182 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0dNq4t9uXVvz-L6rxr:-P0h1qz73CsULAxYJwbv:b-s8q1mr) ([raw file](https://github.com/sillsdev/web-xforge/blob/ffe2864f04f62357d5e530398a99ea511d0b65ae/src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts#L182)):*
> I think that what is meant here is that Obadiah _has chapter content in each chapter in the source_, and so that is why it is not in the list of partial books. If so, could you reword this sentence so it does not use "draft"?

No, that's not correct. Genesis has some content in the target, so the option to draft part of it is present. Obadiah would not get that option due to being a short book. Also, if the target book is blank, we don't offer to draft only part of it. 

I've replaced it with Daniel which will get the chapter input.
___
*[`src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts` line 312 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0dRyT15aLmKNoz30Im:-P0h2WVF3RprHApIHWXx:b-wks030) ([raw file](https://github.com/sillsdev/web-xforge/blob/ffe2864f04f62357d5e530398a99ea511d0b65ae/src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts#L312)):*
<details><summary><i>Previously, marksvc wrote…</i></summary><blockquote>

>  * Choose formatting options, ... apply its first drafted chapter ...
>  * Because that chapter then has content, the import wizard's overwrite confirmation step is guaranteed to appear in
>  * importDraft.
> ...
> await user.click(page.getByRole('button', { name: 'Overwrite chapter' }));

These don't seem to fit together. 

The code line makes me think the project already has content and so we _will_ get an Overwrite prompt that we need to click. And that's okay. But maybe the function comment could use some wording improvement then. Presumably we are benefiting from testing the apply-one-chapter capability, and getting the screenshot.
</blockquote></details>

Yes, it really is more about exercising the chapter import. 

Updated to handle both when the chapter is blank, and when it has content (after the first run it will always have content unless/until the project is modified to remove it).

**Review by marksvc — APPROVED** [review]
> @marksvc reviewed 1 file and all commit messages, made 3 comments, and resolved 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/4082)* status: all files reviewed, 4 unresolved discussions (waiting on Nateowami).

___
*[`src/SIL.XForge.Scripture/ClientApp/e2e/test_characterization.json` line 2 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0dJJpS2h2U3rd28asL:-P0hGkRZEPOMyFTxm8I4:b-dsqo9z) ([raw file](https://github.com/sillsdev/web-xforge/blob/ffe2864f04f62357d5e530398a99ea511d0b65ae/src/SIL.XForge.Scripture/ClientApp/e2e/test_characterization.json#L2)):*
<details><summary><i>Previously, Nateowami wrote…</i></summary><blockquote>

It really serves two purposes:
1. Force us to run a test enough times that we've proven it's somewhat reliable. In practice this is probably the more valuable purpose, even though it wasn't the originally intended purpose. It has definitely stopped me from pushing a test I thought was good.
2. Re-run tests on CI when they fail until it's improbable that they failed for a reason that was present when they were originally characterized. What that means is that when a test fails it's nearly always due to a factor not present when the test was being characterized (machine performance, network speed, code changes, etc). Of course there are numerous ways this could happen without the application having had a true regression.

I've re-characterized generate_draft again.
</blockquote></details>

Do you not find that the E2E tests fail now and then and just need re-run? 
I haven't tried to run them locally for a while because when I have, I can't get the whole suite to run without a failure, and so it hasn't been helpful to tell me if the code is working or broken.

I wonder if it would be helpful to run the characterization on the CI.
___
*[`src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts` line 182 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0dNq4t9uXVvz-L6rxr:-P0hFKuBEeyvL8Y3KbDd:b-7cst3n) ([raw file](https://github.com/sillsdev/web-xforge/blob/ffe2864f04f62357d5e530398a99ea511d0b65ae/src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts#L182)):*
> has 12 chapters, the minimum for the wizard to offer partial drafting,

Thank you for the explanations and comments.
___
*[`src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts` line 200 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/4082#-P0hEFhI1pFY5E-YqQaP:-P0hEFhI1pFY5E-YqQaQ:byfm2i5) ([raw file](https://github.com/sillsdev/web-xforge/blob/bcdcc8d42c3c9c7bd18bc15789c7a45f3988499d/src/SIL.XForge.Scripture/ClientApp/e2e/workflows/generate-draft.ts#L200)):*
I'm surprised for the E2E test to account for these different situations. I would have thought it would *cause* a situation and *assert* what happens.

> case after an earlier run of this test imported the whole book

But from a couple comments I see that this test is running twice, and it looks like with different results each time. 

I think the test would be stronger to cause and assert rather than detect like this. With the detect, the test might not fail if the two runs end up being the same situation instead of experiencing the different situations. Though it might be that if I studied where this test is run from, I would see that it is asserting the different situations.

**Comment by marksvc** [comment]
I approve but am leaving the PR so my comments are not lost :)

**Comment by marksvc** [comment]
Thanks for your persistence with this!
